src/Datadog.Trace/DisposableObjectHandle.cs

```diff
@@ -0,0 +1,38 @@
+#if NET45
+using System;
+using System.Runtime.Remoting;
+using System.Runtime.Remoting.Lifetime;
+
+namespace Datadog.Trace
+{
+    // Create a wrapper around ObjectHandle to enable a Sponsor for
+    // objects stored in the CallContext until the host AppDomain
+    // no longer needs them.
+    // This issue was raised in the Serilog library here: https://github.com/serilog/serilog/issues/987
+    // This solution was borrowed from the corresponding fix in the following PR: https://github.com/serilog/serilog/pull/992
+    internal sealed class DisposableObjectHandle : ObjectHandle, IDisposable
+    {
+        private static readonly ISponsor LifeTimeSponsor = new ClientSponsor();
+
+        public DisposableObjectHandle(object o)
+            : base(o)
+        {
+        }
+
+        public override object InitializeLifetimeService()
+        {
+            var lease = (ILease)base.InitializeLifetimeService();
+            lease?.Register(LifeTimeSponsor);
+            return lease;
+        }
+
+        public void Dispose()
+        {
+            if (GetLifetimeService() is ILease lease)
+            {
+                lease.Unregister(LifeTimeSponsor);
```

Comment:
If this Dispose idempotent? The way it is used in Set(..) it is possible that Dispose is called more than once. If lease.Unregister is not OK to be called multiple times, we need to protect against it.

Reply:
There is no contract stating whether or not this can be called multiple times, but I've tested that calling lease.Unregister multiple times is OK to do. Is this acceptable or would you prefer to be protective anyways?

Reply:
I protected it just in case 😄 in 60708ee7f7caa3ca6f2d8d1bd0fb2fa663c93144